File: By-Python/leetcode/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROW, COL = len(board), len(board[0])
        goal = list(word)
        visit = [[False] * len(board[0]) for _ in range(len(board))]
        
        for row in range(ROW):
            for col in range(COL):
                if board[row][col] == goal[0]:
                    visit[row][col] = True # visit
                    if self.findWord(visit, board, row, col, 1, goal):
                        return True
                    visit[row][col] = False # recover
                    
        return False
    
    def findWord(self, visit, board, row, col, idx, goal):
        ROW, COL = len(board), len(board[0])
        dx = [0,-1,0,1]
        dy = [1,0,-1,0]
        
        # found word
        if idx == len(goal):
            return True
        
        for d in range(4):
            nx = row+dx[d]
            ny = col+dy[d]
            # next location exists in range
            if 0 <= nx < ROW and 0 <= ny < COL:
                # next location is valid character && visit not yet
                if board[nx][ny] == goal[idx] and not visit[nx][ny]:
                    visit[nx][ny] = True # visit
                    if self.findWord(visit, board, nx, ny, idx+1, goal): 
                        return True
                    visit[nx][ny] = False # recover
                
        return False 

File: By-Swift/By-Swift/Leetcode/word-search.swift
class Solution {
    func exist(_ board: [[Character]], _ word: String) -> Bool {
        let R = board.count
        let C = board[0].count
        var visit = Array(repeating: Array(repeating: false, count: C), count: R)
        let word_lst = Array(word)
        
        for i in 0..<R {
            for j in 0..<C {
                if board[i][j] == word_lst[0] {
                    visit[i][j] = true
                    if findWord(board, &visit, word_lst, 1, (i,j)) {
                        return true
                    }
                    visit[i][j] = false
                }
            }
        }
        return false
    }
    
    func findWord(_ board: [[Character]], _ visit: inout [[Bool]], _ word: [Character], _ idx: Int, _ loc: (Int, Int)) -> Bool {
        let R = board.count
        let C = board[0].count
        let dx = [0,1,0,-1]
        let dy = [1,0,-1,0]
        
        if idx == word.count {
            return true
        }
        
        for d in 0..<4 {
            let nx = loc.0 + dx[d]
            let ny = loc.1 + dy[d]
            if (nx >= 0 && nx < R) && (ny >= 0 && ny < C) && !visit[nx][ny] {
                if board[nx][ny] == word[idx] {
                    visit[nx][ny] = true
                    if findWord(board, &visit, word, idx+1, (nx,ny)) {
                        return true
                    }
                    visit[nx][ny] = false
                }
            }
        }
        return false
    }
}